### ISSUE TITLE
Sort out the interaction between scopes of variables

### ISSUE THREAD

**Issue by eponier** [jasmin-lang/jasmin#143]
I create this issue to have something written down.

In pretyping, we have 3 kinds of variables : param, global and local, and it seems to me that the interactions between these scopes are not so clear for the user. For instance,
```
u64 g = 2;
u64[g] g2 = {1};
```
prints the error ``typing error: unknown variable: `g'`` (cf. #63)
while
```
param g = 3;
u64 g = 2;
u64[g] g2 = {1,2,3};
```
will succeed since the size of `g2` is the one coming from the `param` variable, not the `global` one.

In the future, it would be good to have something more intuitive to the user:
- clearer error messages
- do we want variables to shadow other variables, or do we want a clash with an error?

**Comment by eponier** [comment]
After looking at the code with @bgregoir, here are some ideas that we had:
- stop having params and locals on the one hand and globals on the other hand -> just one map
- use long names for variables, so that there is a conflict between two variables defined at the same place, but not with a variable coming from an imported file

**Comment by vbgl** [comment]
This seems to be fixed already (at least in version 2023.06.4):
-  param and globals share the same name-space, so the global `g` actually shadows the parameter with the same name and the use of `g` in `u64[g]` unambiguously refers to the closest declaration;
- shadowing raises a warning.
